sklearn/tests/test_naive_bayes.py

```diff
@@ -945,3 +952,678 @@ def test_n_features_deprecation(Estimator):
+    )
+    assert_array_almost_equal(
+        clf1.predict_log_proba(X2), clf2.predict_log_proba(X2[:, [2, 0, 1, 3, 4]]), 8
+    )
+    assert_array_almost_equal(clf1.predict(X2), clf2.predict(X2[:, [2, 0, 1, 3, 4]]), 8)
+
+
+def test_cwnb_estimators_pandas():
+    pd = pytest.importorskip("pandas")
+    Xdf = pd.DataFrame(data=X, columns=["col0", "col1"])
+    ydf = pd.DataFrame({"target": y})
+
+    # Subestimators spec: cols can be lists of int or lists of str, if DataFrame
+    clf1 = ColumnwiseNB(
+        nb_estimators=[("g1", GaussianNB(), [1]), ("g2", GaussianNB(), [0, 1])]
+    )
+    clf2 = ColumnwiseNB(
+        nb_estimators=[
+            ("g1", GaussianNB(), ["col1"]),
+            ("g2", GaussianNB(), ["col0", "col1"]),
+        ]
+    )
+    clf1.fit(X, y)
+    clf2.fit(Xdf, y)
+    assert_array_almost_equal(clf1.predict_log_proba(X), clf2.predict_log_proba(Xdf), 8)
+    msg = "A column-vector y was passed when a 1d array was expected"
+    with pytest.warns(DataConversionWarning, match=msg):
+        clf2.fit(Xdf, ydf)
+        assert_array_almost_equal(
+            clf1.predict_log_proba(X), clf2.predict_log_proba(Xdf), 8
+        )
+
+    # Subestimators spec: empty cols have the same effect as an absent estimator
+    # when callable columns produce the empty set.
+    select_none = make_column_selector(pattern="qwerasdf")
+    clf1 = ColumnwiseNB(
+        nb_estimators=[
+            ("g1", GaussianNB(), [1]),
+            ("g2", GaussianNB(), select_none),
+            ("g3", GaussianNB(), [0, 1]),
+        ]
+    )
+    clf2 = ColumnwiseNB(
+        nb_estimators=[("g1", GaussianNB(), [1]), ("g3", GaussianNB(), [0, 1])]
+    )
+    clf1.fit(Xdf, y)
+    clf2.fit(Xdf, y)
+    assert_array_almost_equal(
+        clf1.predict_log_proba(Xdf), clf2.predict_log_proba(Xdf), 8
+    )
+    # Empty-columns estimators are passed to estimators_ and the numbers match
+    assert len(clf1.nb_estimators) == len(clf1.estimators_) == 3
+    assert len(clf2.nb_estimators) == len(clf2.estimators_) == 2
+    # No cloning of the empty-columns estimators took place:
+    assert id(clf1.nb_estimators[1][1]) == id(clf1.named_estimators_["g2"])
+
+    # Subestimators spec: test callable columns
+    select_int = make_column_selector(dtype_include=np.int_)
+    select_float = make_column_selector(dtype_include=np.float_)
+    Xdf2 = Xdf
+    Xdf2["col3"] = np.exp(Xdf["col0"]) - 0.5 * Xdf["col1"]
+    clf1 = ColumnwiseNB(
+        nb_estimators=[
+            ("g1", GaussianNB(), ["col3"]),
+            ("m1", BernoulliNB(), ["col0", "col1"]),
+        ]
+    )
+    clf2 = ColumnwiseNB(
+        nb_estimators=[
+            ("g1", GaussianNB(), select_float),
+            ("g2", BernoulliNB(), select_int),
+        ]
+    )
+    clf1.fit(Xdf, y)
+    clf2.fit(Xdf, y)
+    assert_array_almost_equal(
+        clf1.predict_log_proba(Xdf), clf2.predict_log_proba(Xdf), 8
+    )
+
+
+def test_cwnb_estimators_1():
```

Comment:
same remark here about splitting. I will allow you to have more explicit names for the test functions

Reply:
> same remark here about splitting. I will allow you to have more explicit names for the test functions

Done. This and other test functions have been split, and better names have been given to many of them.